csharp/src/Ice/Proxy.cs

```diff
@@ -322,76 +325,170 @@ public static class Proxy
+                                                                      oneway,
+                                                                      synchronous,
+                                                                      observer,
+                                                                      progressWrapper,
+                                                                      cancel).ConfigureAwait(false);
 
-                            if (response.ResultType == ResultType.Failure)
+                            if (response.ResultType != ResultType.Failure)
                             {
-                                observer?.RemoteException();
-
-                                // TODO: revisit
-                                // We throw here the 1.1 system exceptions, as they are used for retries
-                                response.ThrowIfSystemException(proxy.Communicator);
+                                return response;
                             }
-
-                            return response;
+                            // retry
                         }
-                        catch (RetryException)
+                        catch (InvalidRequestHandlerException)
                         {
-                            // Clear the proxy's cached request handler if connection caching is enabled
+                            // TODO temporary, this is already fixed in the transport refactoring branch
+                            // Not a retry, we try to send the request using a handler that was no longer valid
                             if (reference.IsConnectionCached)
                             {
-                                proxy.IceReference.ClearRequestHandler(handler!);
+                                reference.ClearRequestHandler(handler!);
+                            }
+                            continue;
+                        }
+                        catch (NoMoreReplicasException)
+                        {
+                            // The last exception asked to retry using a different replica, but we already tried all
+                            // known replicas.
+                            if (lastException != null)
+                            {
+                                throw ExceptionUtil.Throw(lastException);
+                            }
+                            else
+                            {
+                                Debug.Assert(response != null && response.ResultType == ResultType.Failure);
+                                return response;
                             }
                         }
-                        catch (OperationCanceledException)
+                        catch (ConnectionClosedByPeerException ex)
                         {
-                            throw; // Don't retry cancelled operations
+                            // Always retry a gracefully close connection. Don't exclude the connector after a
```

Comment:
I wonder if this should consume a retry count... what is the risk of always retrying this one with no limits? If a server gracefully closes the connection, presumably it won't accept new connection establishment (if it's shutting down) or it will accept the new connection establishment and won't close again immediately the connection.

Reply:
Yes we should, I accidentally break this with my last change to not use try/catch ...